crypto/s2n_hash.c

```diff
@@ -259,169 +301,121 @@ static int s2n_evp_hash_allow_md5_for_fips(struct s2n_hash_state *state)
+    POSIX_GUARD(s2n_digest_allow_md5_for_fips(&state->digest.high_level.evp));
+    if (s2n_use_custom_md5_sha1()) {
+        POSIX_GUARD(s2n_digest_allow_md5_for_fips(&state->digest.high_level.evp_md5_secondary));
+    }
+    return S2N_SUCCESS;
 }
 
 static int s2n_evp_hash_init(struct s2n_hash_state *state, s2n_hash_algorithm alg)
 {
     POSIX_ENSURE_REF(state->digest.high_level.evp.ctx);
-    POSIX_ENSURE_REF(state->digest.high_level.evp_md5_secondary.ctx);
-    switch (alg) {
-    case S2N_HASH_NONE:
-        break;
-    case S2N_HASH_MD5:
-        POSIX_GUARD_OSSL(EVP_DigestInit_ex(state->digest.high_level.evp.ctx, EVP_md5(), NULL), S2N_ERR_HASH_INIT_FAILED);
-        break;
-    case S2N_HASH_SHA1:
-      POSIX_GUARD_OSSL(EVP_DigestInit_ex(state->digest.high_level.evp.ctx, EVP_sha1(), NULL), S2N_ERR_HASH_INIT_FAILED);
-        break;
-    case S2N_HASH_SHA224:
-        POSIX_GUARD_OSSL(EVP_DigestInit_ex(state->digest.high_level.evp.ctx, EVP_sha224(), NULL), S2N_ERR_HASH_INIT_FAILED);
-        break;
-    case S2N_HASH_SHA256:
-        POSIX_GUARD_OSSL(EVP_DigestInit_ex(state->digest.high_level.evp.ctx, EVP_sha256(), NULL), S2N_ERR_HASH_INIT_FAILED);
-        break;
-    case S2N_HASH_SHA384:
-        POSIX_GUARD_OSSL(EVP_DigestInit_ex(state->digest.high_level.evp.ctx, EVP_sha384(), NULL), S2N_ERR_HASH_INIT_FAILED);
-        break;
-    case S2N_HASH_SHA512:
-        POSIX_GUARD_OSSL(EVP_DigestInit_ex(state->digest.high_level.evp.ctx, EVP_sha512(), NULL), S2N_ERR_HASH_INIT_FAILED);
-        break;
-    case S2N_HASH_MD5_SHA1:
-        POSIX_GUARD_OSSL(EVP_DigestInit_ex(state->digest.high_level.evp.ctx, EVP_sha1(), NULL), S2N_ERR_HASH_INIT_FAILED);
-        POSIX_GUARD_OSSL(EVP_DigestInit_ex(state->digest.high_level.evp_md5_secondary.ctx, EVP_md5(), NULL), S2N_ERR_HASH_INIT_FAILED);
-        break;
-    default:
-        POSIX_BAIL(S2N_ERR_HASH_INVALID_ALGORITHM);
-    }
 
     state->alg = alg;
     state->is_ready_for_input = 1;
     state->currently_in_hash = 0;
 
+    if (alg == S2N_HASH_NONE) {
+        return S2N_SUCCESS;
+    }
+
+    if (alg == S2N_HASH_MD5_SHA1 && s2n_use_custom_md5_sha1()) {
+        POSIX_ENSURE_REF(state->digest.high_level.evp_md5_secondary.ctx);
+        POSIX_GUARD_OSSL(EVP_DigestInit_ex(state->digest.high_level.evp.ctx, EVP_sha1(), NULL), S2N_ERR_HASH_INIT_FAILED);
+        POSIX_GUARD_OSSL(EVP_DigestInit_ex(state->digest.high_level.evp_md5_secondary.ctx, EVP_md5(), NULL), S2N_ERR_HASH_INIT_FAILED);
+        return S2N_SUCCESS;
+    }
+
+    POSIX_ENSURE_REF(s2n_hash_alg_to_evp_md(alg));
+    POSIX_GUARD_OSSL(EVP_DigestInit_ex(state->digest.high_level.evp.ctx, s2n_hash_alg_to_evp_md(alg), NULL), S2N_ERR_HASH_INIT_FAILED);
     return S2N_SUCCESS;
 }
 
 static int s2n_evp_hash_update(struct s2n_hash_state *state, const void *data, uint32_t size)
 {
     POSIX_ENSURE(state->is_ready_for_input, S2N_ERR_HASH_NOT_READY);
+    POSIX_ENSURE(size <= (UINT64_MAX - state->currently_in_hash), S2N_ERR_INTEGER_OVERFLOW);
```

Comment:
I can add an ENSURE_REF to every single static method in this file, but arguably it's unnecessary. The common hash validation is handled by the non-static methods you interact with the module through: https://github.com/aws/s2n-tls/blob/main/crypto/s2n_hash.c#L523-L601.

Reply:
And POSIX_ENSURE_LTE returns a S2N_ERR_SAFETY. I just moved this line, but I'm guessing POSIX_ENSURE was used so that the more specific S2N_ERR_INTEGER_OVERFLOW could be returned.

Reply:
I can live with that rational.